addOns/pscanrulesAlpha/src/main/java/org/zaproxy/zap/extension/pscanrulesAlpha/FetchMetadataRequestHeadersScanRule.java

```diff
@@ -0,0 +1,268 @@
+/*
+ * Zed Attack Proxy (ZAP) and its related class files.
+ *
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ *
+ * Copyright 2023 The ZAP Development Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.zaproxy.zap.extension.pscanrulesAlpha;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import org.parosproxy.paros.Constant;
+import org.parosproxy.paros.core.scanner.Alert;
+import org.parosproxy.paros.network.HttpHeaderField;
+import org.parosproxy.paros.network.HttpMessage;
+import org.parosproxy.paros.network.HttpRequestHeader;
+import org.zaproxy.addon.commonlib.CommonAlertTag;
+import org.zaproxy.zap.extension.pscan.PluginPassiveScanner;
+
+public class FetchMetadataRequestHeadersScanRule extends PluginPassiveScanner {
+    private static final String FETCH_METADATA_REQUEST_MESSAGE_PREFIX =
+            "pscanalpha.metadata-request-headers.";
+    private static final int PLUGIN_ID = 90005;
+    private final List<FetchMetaDataRequestHeaders> rules =
+            List.of(
+                    new SecFetchSite(this::newAlert),
+                    new SecFetchMode(this::newAlert),
+                    new SecFetchDest(this::newAlert),
+                    new SecFetchUser(this::newAlert));
+
+    private static final Map<String, String> ALERT_TAGS =
+            CommonAlertTag.toMap(CommonAlertTag.WSTG_V42_SESS_05_CSRF);
+
+    @Override
+    public void scanHttpRequestSend(HttpMessage msg, int id) {
+        rules.forEach(
+                rule ->
+                        rule.build(msg.getRequestHeader(), rule.getHeader(), rule.getValidValues())
+                                .forEach(AlertBuilder::raise));
+    }
+
+    @Override
+    public int getPluginId() {
+        return PLUGIN_ID;
+    }
+
+    @Override
+    public String getName() {
+        return Constant.messages.getString(FETCH_METADATA_REQUEST_MESSAGE_PREFIX + "name");
+    }
+
+    @Override
+    public List<Alert> getExampleAlerts() {
+        List<Alert> alerts = new ArrayList<>();
+
+        alerts.addAll(
+                rules.stream()
+                        .map(s -> s.alert("missing.", "").build())
```

Comment:
This is fragile, better have the rules  return the example alerts rather than add conditions here.

Reply:
Could you please elaborate. I have followed the same approach used for other scan rules like `SiteIsolationScanRule.java`.

Reply:
It's not the same thing, `SiteIsolationScanRule` builds the alerts directly it's not passing keys to create different alerts.

First step would be to add the method `FetchMetaDataRequestHeaders.getExampleAlerts()` and extract the `vulnType` logic, so it can be used by both methods.

Note that the alerts for missing header and invalid values should both have its own alert ref (it should not be per site isolation rule but per alert raised).

Reply:
https://github.com/zaproxy/zap-extensions/pull/4565#discussion_r1246943139

I don't think we want to convey meaning to the alert references, that's just meant to identify an alert type. All other scan rules are using sequential numbers and I think we should do that here as well. @kingthorin @psiinon @ricekot thoughts?

https://github.com/zaproxy/zap-extensions/pull/4565#discussion_r1246948376

Thinking better, extract methods e.g. `createMissingHeaderAlert()` and `createInvalidValueAlert(String)`, and use that in `getExampleAlerts()` and `build(…)`. It still repeats the calls to those methods but less movable parts.

Reply:
I agree they’re simple differentiators, they shouldn’t have “meaning”.

I agree with the extraction idea.

Reply:
added changes

Reply:
And I noticed now that the alert refs were not updated yet.

Reply:
> And I noticed now that the alert refs were not updated yet.

Apologies for missing out on this.

Reply:
No worries.